src/aa_remove_data/pb_utils.py

```diff
@@ -139,49 +182,75 @@ def read_pb(self, filepath: PathLike):
+                )
+                self._total_lines = int(result.stdout.split()[0])
+                first_line = self._restore_newline_chars(f.readline().strip())
+                self.header.ParseFromString(first_line)
+                f.seek(0)
+                self._start_line += 1
+            end_line = min(self._start_line + self._chunk_size, self._total_lines)
+            lines = list(islice(f, self._start_line, end_line))
+            if self._total_lines == end_line:
+                self.read_done = True
+            else:
+                self.chunked = True
+            self._start_line = end_line
+            sample_class = self.get_proto_class()
+            self.samples = [sample_class() for _ in range(len(lines))]
+            for i, sample in enumerate(self.samples):
+                line = self._restore_newline_chars(lines[i].strip())
+                sample.ParseFromString(line)
 
     def write_pb(self, filepath: PathLike):
-        """Write a PB file that is structured in the Archiver Appliance format.
+        """Write a pb file that is structured in the Archiver Appliance format.
         Must have a valid header and list of samples to write.
 
         Args:
             filepath (PathLike): Path to file to be written.
         """
-        header_b = self._replace_newline_chars(self.header.SerializeToString()) + b"\n"
         samples_b = [
             self._replace_newline_chars(sample.SerializeToString()) + b"\n"
             for sample in self.samples
         ]
-        with open(filepath, "wb") as f:
-            f.writelines([header_b] + samples_b)
+        if (
+            filepath not in self._write_started or self.chunked is False
+        ):  # Write header, start new file
+            header_b = (
+                self._replace_newline_chars(self.header.SerializeToString()) + b"\n"
+            )
+            with open(filepath, "wb") as f:
+                f.writelines([header_b] + samples_b)
+            self._write_started.append(filepath)
+        else:  # Add to existing file
+            with open(filepath, "ab") as f:
+                f.writelines(samples_b)
 
 
 def pb_2_txt():
-    """Convert a .pb file to a human-readable .txt file"""
+    """Convert a .pb file to a human-readable .txt file."""
     parser = argparse.ArgumentParser()
     parser.add_argument("pb_filename", type=str)
     parser.add_argument("txt_filename", type=str)
+    parser.add_argument("--chunk", default=10000000)
     args = parser.parse_args()
     pb_file = Path(args.pb_filename)
     txt_file = Path(args.txt_filename)
     if not pb_file.is_file():
         raise FileNotFoundError(f"The file {pb_file} does not exist.")
     if pb_file.suffix != ".pb":
         raise ValueError(f"Invalid file extension: '{pb_file.suffix}'. Expected '.pb'.")
-    pb = PBUtils(pb_file)
-    pb.write_to_txt(txt_file)
+    pb = PBUtils(chunk_size=args.chunk)
+    while pb.read_done is False:
+        pb.read_pb(pb_file)
```

Comment:
The read continues to only write the length equal to the chunk size. What I think is happening here is that the write_to_txt function opens the file with write ("w") which keeps overwriting the file. What we end up with then is the last x lines equal to the chunk size.

From our chats, you may have resolved this already in the remove_data branch. It may just need porting across.